src/shipLogEntries.ts

```diff
@@ -180,5 +181,5 @@ export async function shipLogEntries(event: CloudWatchLogsEvent, context: Contex
     });
     console.log(`Sending ${structuredLogs.length} events from ${logGroup} to ${kinesisStreamName} (with role: ${kinesisStreamRole})`);
     
-    await putKinesisRecords(kinesis, kinesisStreamName, structuredLogs);
+    return await putKinesisRecords(kinesis, kinesisStreamName, structuredLogs);
```

Comment:
The [invocation retries documentation](https://docs.aws.amazon.com/lambda/latest/dg/invocation-retries.html) states that common function errors are _"Your function's code throws an exception or returns an error object"_. Whilst I couldn't find an explicit definition of the _"error object"_ I'm presuming it should have fields `errorType` and `errorMessage` [e.g.](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-exceptions.html)

In particular, it seems that if there are instances of  `FailedRecordCount > 0` in the [`PutRecordsOutput[]`](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecords.html), the lambda won't consider this as an error.

Presuming this is correct (?), should this be resolved by e.g. iterating through the `PutRecordsOutput[]` and raising an exception if `FailedRecordCount > 0` ?

Reply:
Interesting and probably.

At the very least we should log a line saying how many records we had failed to put, but given that this logging doing end up anywhere helpful...

Thinking about it we could directly ship to the ELK stream from this function so we could ship an ERROR level log entry from this function if it fails.

Ultimately though this is a question of how we are informed of issues.

Reply:
Thinking more, given the current mechanisms we have in place, if there are failures we should log stats and then throw. I suspect Error object is anything that extends Error so returning rather than throwing an exception?

Reply:
> Ultimately though this is a question of how we are informed of issues.

Yes - good point. Ok, I won't look to solve this in the current PR, but will create an issue. I guess as part of solving this, we could also think about the case when the Promise is rejected. In that case it's reported as a failed invocation (good!) but perhaps notification needs to be automated - alarms etc.

Reply:
> Thinking more, given the current mechanisms we have in place, if there are failures we should log stats and then throw. I suspect Error object is anything that extends Error so returning rather than throwing an exception?

Yes, this seems like a good starting point. After this change then at least alarms can be configured per account if the respective team deems it appropriate.

Ok, to follow this up in a separate PR ?